### PR TITLE
add control flow handling to DenseLayout pass

### PR DESCRIPTION
**Pull request by ewinston**
### Summary
Add control flow handling to dense layout pass.

### Details and comments
depends on #8763

todo: may still need to update `Target.operation_names_for_qargs` when target object is supplied.

### CODE REVIEW

**Comment by mtreinish** [comment]
FWIW, I'm not sure how much this is actually needed. The cx count and measurement count are only used as weighting factors when trying to estimate how noisy the qubits in a layout are (in the `BackendV1` path only). For control flow circuits trying to adjust that count really doesn't provide much benefit over not doing it because it's just a heuristic to try and avoid super noisy qubits or 2q links being selected in the layout (the weighting factor is just to care more about 2q error vs measurements based on the circuit which is why we excluded it from the v2 path). This error based heuristic is kind of obsolete now anyway with `VF2PostLayout` which does a much better job at noise aware qubit selection post routing so we probably could remove the heuristic from the pass now.

File: qiskit/dagcircuit/dagcircuit.py [hunk 1]
```diff
@@ -1738,12 +1739,19 @@ def nodes_on_wire(self, wire, only_ops=False):
...
         """
-        return self._op_names.copy()
+        op_count = self._op_names.copy()
+        if expand_control_flow:
+            for cfop in self.op_nodes(op=ControlFlowOp):
+                if cfop.name in op_count:
+                    del op_count[cfop.name] 
+                for block in cfop.op.blocks:
+                    op_count = Counter(op_count) + Counter(block.count_ops())
+        return op_count
```
Comment (jakelishman) [inline qiskit/dagcircuit/dagcircuit.py:1754]: I did this (along with a few of the other DAGCircuit methods) in #8763 as well, so we probably only need one.  In #8763 I actually made the default `True`, so this pass would automatically update itself, though the test would be good to keep.

File: test/python/transpiler/test_dense_layout.py [hunk 2]
```diff
@@ -16,14 +16,15 @@
...
-from qiskit import QuantumRegister, QuantumCircuit
+from qiskit import QuantumRegister, QuantumCircuit, ClassicalRegister
 from qiskit.circuit import Parameter
 from qiskit.circuit.library import CXGate, UGate, ECRGate, RZGate
 from qiskit.transpiler import CouplingMap, Target, InstructionProperties, TranspilerError
 from qiskit.transpiler.passes import DenseLayout
 from qiskit.converters import circuit_to_dag
 from qiskit.test import QiskitTestCase
 from qiskit.providers.fake_provider import FakeTokyo
+from qiskit.visualization.gate_map import *
```
Comment (jakelishman) [inline test/python/transpiler/test_dense_layout.py:27]: This seems questionable?

**Review by jakelishman — COMMENTED** [review]
#8763 will automatically upgrade `DenseLayout` to recurse into control-flow when counting up the `cx` and `measure` gates, and that PR is perhaps a bit more logically the right location for such a change.  I get the argument that `DenseLayout`'s heuristics are going to be a bit rubbish for control flow, but I think having the explicit test, and just having them be automatically upgraded is fine - if we want to revisit the heuristic, it's better to do that in a non-control-flow specific PR anyway.

File: test/python/transpiler/test_dense_layout.py (hunk 2 above)
Comment (ewinston) [reply test/python/transpiler/test_dense_layout.py]: debug code. Thanks for spotting. removed in e390d476

**Review by jakelishman — APPROVED** [review]
I've rebased this properly over #8763 now that it's merged.  The principal remaining change is now mostly the test, and since this is just a simple pass, I think this is fine.

**Comment by jakelishman** [comment]
The heuristic here is really pretty meaningless (and clearly intended for backends that have cx as their only 2+q gate), and the `Target` path doesn't even attempt to put a number on it.  I'm not bothered at all about doing more work on the heuristic, since it's already not really doing anything.

**Pull request merged**